Documentation guide: Solve a system of equations numerically (#23853)
sympy/sympy: doc/src/guides/solving/solve-numerically.md

- starting values `(-1, 1)`

```py
>>> from sympy import Symbol, nsolve
>>> x1 = Symbol('x1')
>>> x2 = Symbol('x2')
>>> f1 = 3 * x1**2 - 2 * x2**2 - 1
>>> f2 = x1**2 - 2 * x1 + x2**2 + 2 * x2 - 8
>>> print(nsolve((f1, f2), (x1, x2), (-1, 1)))
Matrix([[-1.19287309935246], [1.27844411169911]])
```

### Solve overdetermined systems

In an overdetermined system, there are more equations than unknowns. In general, this means that there is no exact solution, but there will be an approximate solution that is the closest fit to the equations. Overdetermined systems are common when the data is noisy, and the approximate solution is useful rather than giving up because there is no exact solution.

*Thought this would work but it throws an error*

from sympy import Symbol, nsolve

x1 = Symbol('x1')

x2 = Symbol('x2')

f1 = 3 * x1**2 - 2 * x2**2 - 1

f2 = x1**2 - 2 * x1 + x2**2 + 2 * x2 - 8

f3 = x1 + x2

print(nsolve((f1, f2, f3), (x1, x2), (-1, 1)))

### Use SciPy on a lambda function *what is the advantage of this method?*

You can *description*

```py
>>> from sympy.abc import x
>>> from sympy.utilities.lambdify import implemented_function
>>> from sympy import lambdify

> >>> from scipy import optimize

Author: This doctest failed because SciPy wasn't available for import. How do I add that module? Or is using SciPy a bad idea?

Reviewer: You need to add it to `_get_doctest_blacklist` in `sympy/testing/runtests.py` so that it only gets run when scipy is installed.

Author: I tried to follow the pattern there, if this isn't the correct technique please let me know:

```
    if import_module("scipy") is None:
        # throws ModuleNotFoundError when scipy not installed
        blacklist.extend(
            ["doc/src/guides/solving/solve-numerically.md",]
        )
```

And should I add scipy to `doc/requirements.txt` so scipy will be available for import? I would think we don't want to permanently prevent doctests on the new page `doc/src/guides/solving/solve-numerically.md`...

Reviewer: > And should I add scipy to doc/requirements.txt so scipy will be available for import? 

No that file is just for things needed to build the documentation. 

> I would think we don't want to permanently prevent doctests on the new page doc/src/guides/solving/solve-numerically.md...

If you did the logic correctly it will skip the file when scipy isn't installed but run the doctests when it is. The optional dependencies build on CI has scipy installed so you can check if it is running the tests there.

Author: Understand. I checked tests such as `doctests (3.8, false)` and it includes the [new page passing doctest](https://github.com/sympy/sympy/runs/7587802527?check_suite_focus=true#step:6:531), so seems like my addition to `_get_doctest_blacklist` in `sympy/testing/runtests.py`.

# Solve one or a system of equations numerically

Use SymPy to numerically solve a system of one or more equations. For example,
numerically solving $\cos(x) = x $ returns $ x \approx 0.739085133215161$.

Solving numerically is useful if:
- You only need a numeric solution, not a symbolic one
- A closed-form solution is not available or is overly complicated

{func}`~.solve` and {func}`~.solveset` will not try to find a numeric solution,
only a mathematically-exact symbolic solution. So if you want a numeric
solution, use {func}`~.nsolve`.

Alternatives to consider:
- [NumPy](https://numpy.org/doc/stable/reference/generated/numpy.linalg.solve.html?highlight=solve#numpy.linalg.solve)
and
[SciPy](https://docs.scipy.org/doc/scipy/reference/generated/scipy.linalg.solve.html#scipy.linalg.solve)

> can each solve a system of linear scalar equations

Author: Reminding myself to use intersphinx for these references

the interval `(-10, 0)` ensures that the root `-1` is found:

```py
>>> from sympy import nsolve
>>> from sympy.abc import x
>>> nsolve(x**2 - 1, (-10, 0), solver='bisect')
-1.00000000000000
```

## Solve a system of equations numerically

### Solve multidimensional functions

To solve multidimensional functions, supply a tuple of
- functions `(f1, f2)`
- variables to solve for `(x1, x2)`
- starting values `(-1, 1)`

```py
>>> from sympy import Symbol, nsolve
>>> x1 = Symbol('x1')
>>> x2 = Symbol('x2')
>>> f1 = 3 * x1**2 - 2 * x2**2 - 1
>>> f2 = x1**2 - 2 * x1 + x2**2 + 2 * x2 - 8
>>> print(nsolve((f1, f2), (x1, x2), (-1, 1)))
Matrix([[-1.19287309935246], [1.27844411169911]])
```

### Use SciPy on a lambda function *what is the advantage of this method?*

You can *description*

```py
>>> from sympy.abc import x
>>> from sympy.utilities.lambdify import implemented_function
>>> from sympy import lambdify
>>> from scipy import optimize
>>> f = implemented_function('f', lambda x: x**3 - 1)
>>> lam_f = lambdify(x, f(x))
>>> sol = optimize.root_scalar(lam_f, bracket=[0, 2], method='brentq')
>>> sol.root
1.0
```

## Use the solution result

### Substitute the result into an expression

> You can use [`subs`](sympy.core.basic.Basic.subs) to substitute numerical values

Reviewer 2: The best way to do this is to do `expr.n(subs={x: xvalue})` -- or use `evalf` instead of `n` -- per evalf note:
```
Notes
=====

When Floats are naively substituted into an expression,
precision errors may adversely affect the result. For example,
adding 1e16 (a Float) to 1 will truncate to 1e16; if 1e16 is
then subtracted, the result will be 0.
That is exactly what happens in the following:

>>> from sympy.abc import x, y, z
>>> values = {x: 1e16, y: 1, z: 1e16}
>>> (x + y - z).subs(values)
0

Using the subs argument for evalf is the accurate way to
evaluate such an expression:

>>> (x + y - z).evalf(subs=values)
1.00000000000000
```
Since you are getting numerical results and are going to use them in some way, this would be a good habit to form early.

>>> derivative.subs(x, x_value) # Calculate the derivative where the function value is zero
-1.67361202918321
```

or leave some symbols as variables:

```py
>>> from sympy import cos, nsolve, Symbol, diff
>>> x = Symbol('x')
>>> f = cos(x) - x
>>> x_value = nsolve(f, x, 1); x_value
0.739085133215161
>>> y = Symbol('y')
>>> g = x * y**2
>>> g.subs(x, x_value)
0.739085133215161*y**2
>>> derivative = diff(g, y)
>>> derivative.subs(x, x_value)
1.47817026643032*y
```

## Tradeoffs

### Do not use `verify` for functions which are very steep near the root

For functions which are very steep near the root, the verification of the
solution may fail. In this case you should use the flag `verify=False` and
independently verify the solution.

```py
>>> from sympy import cos, cosh, nsolve, Symbol
>>> x = Symbol('x')
>>> f = cos(x)*cosh(x) - 1
>>> nsolve(f, 3.14*100)
Traceback (most recent call last):
...
ValueError: Could not find root within given tolerance. (1.39267e+230 > 2.1684e-19)
>>> ans = nsolve(f, 3.14*100, verify=False); ans
312.588469032184
>>> f.subs(x, ans).n(2)
2.1e+121
>>> (f/f.diff(x)).subs(x, ans).n(2)
7.4e-15
```

## Not all equations can be solved

> {func}`~.nsolve` is a numerical solving function, so it is often the solution to

Reviewer 2: ```suggestion
{func}`~.nsolve` is a numerical solving function, so it can often provide a solution for
```

equations which cannot be solved algebraically.

### Equations with no solution

Some equations have no solution, in which case SymPy may return an empty set.

> For example, the equation $x - 7 - x - 2 = 0$ reduces to $-9 = 0$, which has no

Reviewer 2: In this case the equation trivially simplified to a number and `nsolve` doesn't accept that. But `sqrt(x) + 2` might be a better example since there is no solution for that.

Author: Thanks for the suggestion. Please help me understand--while both SymPy and [Wolfram Alpha](https://www.wolframalpha.com/input?i2d=true&i=Sqrt%5Bx%5D%2B2%3D0) agree with you, I thought the square root of 4 could be either 2 or -2, and if was chosen as -2, then 4 would be a solution of the equation you give?

Reviewer: SymPy assumes that sqrt is only the principal square root. 

A more straightforward example might be `exp(x) = 0`, which doesn't have any (finite) solutions, even in a complex setting.

Author: Thanks, I went with `exp(x) = 0` for conceptual simplicity. A wrinkle is that it requires greater than default precision to return the correct response that there are no solutions. That led me to introduce `prec` earlier in the page, which I think could be valuable to many visitors.

>>> x_value = nsolve(f, x, 1); x_value
0.739085133215161
>>> y = Symbol('y')
>>> z = Symbol('z')
>>> g = x * y**2
>>> values = {x: x_value, y: 1}
>>> (x + y - z).evalf(subs=values)
1.73908513321516 - z
```

## Not all Equations Can be Solved

{func}`~.nsolve` is a numerical solving function, so it can often provide a
solution for equations which cannot be solved algebraically.

### Equations With no Solution

Some equations have no solution, in which case SymPy may return an error. For
example, the equation $e^x = 0$ (`exp(x)` in SymPy) has no solution:

```py
>>> from sympy import nsolve, exp
>>> from sympy.abc import x
>>> nsolve(exp(x), x, 1, prec=20)
Traceback (most recent call last):
...
ValueError: Could not find root within given tolerance. (5.4877893607115270300540019e-18 > 1.6543612251060553497428174e-24)
Try another starting point or tweak arguments.
```

## Report a Problem

If you find a problem with {func}`~.nsolve`, please post it on the [mailing
list](https://groups.google.com/g/sympy), or open an issue on [SymPy's GitHub
page](https://github.com/sympy/sympy/issues). Until the issue is resolved, you
can use NumPy's {external:func}`numpy.linalg.solve` or SciPy's

> {external:func}`scipy.linalg.solve`.

Reviewer 3: The proper analogue of `nsolve` is SciPy's `fsolve` because `nsolve` is not specifically for linear equations.

>>> x2 = Symbol('x2')
>>> f1 = 3 * x1**2 - 2 * x2**2 - 1
>>> f2 = x1**2 - 2 * x1 + x2**2 + 2 * x2 - 8
>>> print(nsolve((f1, f2), (x1, x2), (-1, 1)))
Matrix([[-1.19287309935246], [1.27844411169911]])
```

### Increase Precision of the Solution

You can increase the precision of the solution using `prec`:

```py
>>> from sympy import Symbol, nsolve
>>> x1 = Symbol('x1')
>>> x2 = Symbol('x2')
>>> f1 = 3 * x1**2 - 2 * x2**2 - 1
>>> f2 = x1**2 - 2 * x1 + x2**2 + 2 * x2 - 8
>>> print(nsolve((f1, f2), (x1, x2), (-1, 1), prec=25))
Matrix([[-1.192873099352460791205211], [1.278444111699106966687122]])
```

### Create a Function That Can Be Solved With SciPy

As noted above, SymPy focuses on symbolic computation and is not optimized for
numerical calculations. A recommend use case for SymPy in a workflow is as given
below:
- use SymPy to generate (by symbolically simplifying or solving an equation) the
  mathematical expression
- use a numerical library such as SciPy to generate numerical solutions

```py
>>> from sympy import simplify, cos, sin
>>> from sympy.abc import x, y
>>> expr = cos(x * (x + x**2)/(x*sin(y)**2 + x*cos(y)**2 + x))
>>> simplify(expr)
cos(x*(x + 1)/2)

> >>> from sympy.utilities.lambdify import implemented_function

Reviewer: I don't see the point of using `implemented_function` here. I guess this is coming from its use in the `lambdify` docstring, but I don't really think it's necessary there either. `implemented_function` is not something that I think most people really need to use.

>>> from sympy import lambdify
>>> f = implemented_function('f', lambda x: cos(x*(x + 1)/2))
>>> lam_f = lambdify(x, f(x))
>>> from scipy import optimize
>>> sol = optimize.root_scalar(lam_f, bracket=[0, 2], method='brentq')
>>> sol.root
1.3416277185114782
```

## Use the Solution Result

### Substitute the Result Into an Expression

The best practice is to use {func}`~sympy.core.evalf` to to substitute numerical
values into expressions. Using [`subs`](sympy.core.basic.Basic.subs) can give
incorrect result due to precision errors:

```py
>>> from sympy import cos, nsolve, Symbol
>>> x = Symbol('x')
>>> f = cos(x) - x
>>> x_value = nsolve(f, x, 1); x_value
0.739085133215161
>>> y = Symbol('y')
>>> z = Symbol('z')
>>> values = {x: x_value, y: 1e16, z: 1e16}
>>> (x + y - z).evalf(subs=values) # Gives correct result
0.739085133215161

> >>> (x + y - z).subs(values) # Gives incorrect result due to precision errors

Reviewer: This example (and the one below) seem a bit odd because you are substituting the solution from an equation into another expression (`x + y - z`) and it's not clear why you would do that. In fact, the issue here has nothing to do with the value 0.73... but rather from the cancellation from 1e16 and -1e16.

Maybe a more motivated example would be showing substituting the solved value back into the equation. This has the same fundamental issue because of the subtraction in `cos(x) - x`.

```py
>>> f.subs(x, x_value)
0
>>> f.evalf(subs={x: x_value})
-5.12757857962640e-17
```

Here the second answer is more correct, because the `x_value` is not an exact solution to the equation, so substituting it should not give exactly 0.

Author: That [example was suggested](https://github.com/sympy/sympy/pull/23853#discussion_r933978731) by @smichr from the [EvalfMixin](https://docs.sympy.org/latest/modules/core.html?highlight=subs#sympy.core.evalf.EvalfMixin) docstring, which notes "Using the subs argument for evalf is the accurate way to evaluate such an expression."

It doesn't strike me as odd to substitute the solution from an equation into another expressions--it seems like that's something you'd do when you have one equation that you can use to solve for `x`, and then want to use that value of `x` to find the value of another expression.

If desired, I could return to the docstring's formulation where the value (that we don't want to get wiped out by a precision error) comes not from the solution to another equation, but is just given as a numerical value: `values = {x: 1e16, y: 1, z: 1e16}`.

Reviewer: My problem is that the example feels very contrived. Substituting the solution into the original equation is a great example of where you are very likely to get loss of precision issues, almost by definition. And it doesn't actually require you to have large values like 1e16 for that to occur. 

Really I would prefer if this wasn't necessary and it just worked correctly automatically with normal `subs`, but that requires some improvements to `subs`.

Author: Can you outline an example that wouldn't feel contrived? I think the point here is to alert visitors to the gotcha by demonstrating it and how to avoid it.

Certainly it would be best if this gotcha didn't exist, so this workaround wasn't necessary.

Reviewer: I think substituting into the original equation isn't contrived, and it's also one of the more likely situations the gotcha would arise, since you're almost guaranteed to have loss of significance if the original expression has multiple terms in it.

Author: Substituting into the original equation doesn't seem to exhibit the gotcha--it gives identically zero, as you would expect:

```py
>>> from sympy import cos, nsolve, Symbol
>>> x = Symbol('x')
>>> f = cos(x) - x
>>> x_value = nsolve(f, x, 1)
>>> x_value
0.739085133215161
>>> f.subs(x, x_value)
0
```

Reviewer: Maybe it's too subtle, but the point is that the expression *shouldn't* be identically zero. The value 0.739085133215161 is only an approximation to the true solution, which is infinitely many digits long, so substituting it into the expression should produce a value which is close to 0, but not exactly equal to it. The reason 0 is returned is because the naive substitution isn't smart enough to increase the precision of the operation to produce the actual value, which you can see from the `evalf(subs)` example above is around 1e-17. More specifically, in the default precision (15 digits), cos(0.739085133215161) just happens to produce exactly 0.739085133215161, so if you subtract them, you get 0. 

Another way to look at this is to say that the solution `0.739085133215161` is accurate to within 1e-16, since when plugged back into the original equation you get `-5.12757857962640e-17`. This is expected since the default precision in `nsolve` is 15, meaning the answer should be accurate to 15 digits (within 1e-15). Similarly, if you increase the precision requested, the result is accurate to within that precision:

```py
>>> x_value = nsolve(cos(x) - x, x, 1, prec=40)
>>> (cos(x) - x).evalf(subs={x: x_value})
-8.91241325272802e-42
```

(of course, it's not actually necessary for the user to check this, as `nsolve` does so automatically, but that could be seen as a motivating reason why you would do it)

0
```

You can also leave some symbols as variables:

```py
>>> from sympy import cos, nsolve, Symbol
>>> x = Symbol('x')
>>> f = cos(x) - x
>>> x_value = nsolve(f, x, 1); x_value
0.739085133215161
>>> y = Symbol('y')
>>> z = Symbol('z')
>>> g = x * y**2
>>> values = {x: x_value, y: 1}

> >>> (x + y - z).evalf(subs=values)

Reviewer: I thought this was broken https://github.com/sympy/sympy/issues/6974. Maybe that issue is fixed now?

{func}`~.nsolve` is a numerical solving function, so it can often provide a
solution for equations which cannot be solved algebraically.

### Equations with no solution

Some equations have no solution, in which case SymPy may return an error. For
example, the equation $exp(x) = 0$ has no solution:

```py
>>> from sympy import nsolve, exp
>>> from sympy.abc import x
>>> nsolve(exp(x), x, 1, prec=20)
Traceback (most recent call last):
...
ValueError: Could not find root within given tolerance. (5.4877893607115270300540019e-18 > 1.6543612251060553497428174e-24)
Try another starting point or tweak arguments.

> ```

Reviewer 2: On the other hand, this function gets arbitrarily close to 0 and the default solver will report a false solution because at that value the function is very small:
```python
>>> nsolve(exp(x), x, -1)
-21.8720030830002
```
And currently the bisect routine does not check the sign before checking the value and does not flag this case:
```python
>>> nsolve(exp(x),(-22,-21), solver='bisect')
-22.0000000000000
>>> sign(exp(-22)*exp(-21))
1
>>> nsolve(x**2-1,x,(2,3),solver='bisect')
Traceback (most recent call last):
...
ValueError: Could not find root within given tolerance. (9.0000000000000000052 > 2.16840434497100886801e-19)
Try another starting point or tweak arguments.
```

Author: Yeah, haven't found a great example for this--each has some wrinkle to it.

# Solve a system of equations numerically

Use SymPy to numerically solve a system of one or more equations. For example,
numerically solving $\cos(x) = x $ returns $ x \approx 0.739085133215161$.

Alternatives to consider:
- [NumPy](https://numpy.org/doc/stable/reference/generated/numpy.linalg.solve.html?highlight=solve#numpy.linalg.solve)
and
[SciPy](https://docs.scipy.org/doc/scipy/reference/generated/scipy.linalg.solve.html#scipy.linalg.solve)
can each solve a system of linear scalar equations
- To solve an underdetermined system of linear equations, convert them to
  matrices and solve using [`gauss_jordan_solve`](gauss_jordan_solve).

Here is a simple example of numerically solving one equation:

```py
>>> from sympy import cos, nsolve, Symbol
>>> x = Symbol('x')
>>> nsolve(cos(x) - x, x, 1)
0.739085133215161
```

{func}`~.nsolve` calls, and can pass parameters to,

> [mpmath.findroot](https://mpmath.org/doc/current/calculus/optimization.html#root-finding-findroot).

Reviewer: We should use intersphinx (see how it is done currently in conf.py for matplotlib, and then reference it with `{external}`).

0
>>> derivative = diff(f, x)
>>> derivative.subs(x, x_value) # Calculate the derivative where the function value is zero
-1.67361202918321
```

or leave some symbols as variables:

```py
>>> from sympy import cos, nsolve, Symbol, diff
>>> x = Symbol('x')
>>> f = cos(x) - x
>>> x_value = nsolve(f, x, 1); x_value
0.739085133215161
>>> y = Symbol('y')
>>> g = x * y**2
>>> g.subs(x, x_value)
0.739085133215161*y**2
>>> derivative = diff(g, y)
>>> derivative.subs(x, x_value)
1.47817026643032*y
```

## Tradeoffs

### Do not use `verify` for functions which are very steep near the root

For functions which are very steep near the root, the verification of the
solution may fail. In this case you should use the flag `verify=False` and
independently verify the solution.

```py
>>> from sympy import cos, cosh, nsolve, Symbol
>>> x = Symbol('x')
>>> f = cos(x)*cosh(x) - 1
>>> nsolve(f, 3.14*100)
Traceback (most recent call last):
...
ValueError: Could not find root within given tolerance. (1.39267e+230 > 2.1684e-19)
>>> ans = nsolve(f, 3.14*100, verify=False); ans
312.588469032184
>>> f.subs(x, ans).n(2)
2.1e+121
>>> (f/f.diff(x)).subs(x, ans).n(2)

> 7.4e-15

Reviewer: This is from the docstring. Do other people think this is worth discussing here, or is this too advanced for this guide? It's a bit subtle IMO, and really could use better explanation of what's going on.

Reviewer 2: There are lots of numerical tricks to getting an equation in a well posed form for solving. Some general things that might be mentioned -- for which SymPy's symbolics can help -- would be good:
* working with the numerator: `n, d = eq.as_numer_denom(); nsolve(n, ...)`
* dividing by the derivative: `fd = eq/eq.diff(); nsolve(fd...)`
* removing large contants/common factors: `_eq = factor_terms(eq).as_coeff_Mul()[1]; nsolve(_eq...)`

But knowing how to get a root using `verify=False` when you know one is there but the routine is having troubles is a useful hint. But it can also be worked around (sometimes) by using some of the above methods:
```python
>>> f
cos(x)*cosh(x) - 1
>>> nsolve(f/f.diff(),312)
312.588469032184
```

Reviewer: It would be good to give a little mathematical reasoning for these tricks. Without that, your first and second bullet points appear to contradict one another. 

The first point matters, if I understand correctly, if you get ZeroDivisionError. But it's not the only way that zero division can occur, so in some cases more advanced rewriting may be necessary.

The second bullet point is relevant when the root in question is a double root. Here's a simple example:

```
>>> nsolve(cos(x) - 1, 3)
-515.221191947442
>>> nsolve((cos(x) - 1)/-sin(x), 3)
6.28318530717931
```

The solution is correct (it's -164π), but it's not the closest one to the starting point. It's also impossible to use the bisection solver unless the root has odd multiplicity (it has to alternate signs at the endpoints of the interval).

```
>>> nsolve(cos(x) - 1, [1, 13], solver='bisect')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/aaronmeurer/Documents/Python/sympy/sympy/./sympy/utilities/decorator.py", line 88, in func_wrapper
    return func(*args, **kwargs)
  File "/Users/aaronmeurer/Documents/Python/sympy/sympy/./sympy/solvers/solvers.py", line 2979, in nsolve
  File "/Users/aaronmeurer/anaconda3/lib/python3.9/site-packages/mpmath/calculus/optimization.py", line 985, in findroot
    raise ValueError('Could not find root within given tolerance. '
ValueError: Could not find root within given tolerance. (0.211321969990149371762 > 2.16840434497100886801e-19)
Try another starting point or tweak arguments.
>>> nsolve((cos(x) - 1)/-sin(x), [1, 13], solver='bisect')
6.28318530717661
```

OTOH, this doesn't seem to always work

```
>>> nsolve(expand((x - 1)**4), 0)
0.998067087259351
>>> nsolve(expand((x - 1)**4)/diff((x - 1)**4, x).expand(), 0)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/aaronmeurer/Documents/Python/sympy/sympy/./sympy/utilities/decorator.py", line 88, in func_wrapper
    return func(*args, **kwargs)
  File "/Users/aaronmeurer/Documents/Python/sympy/sympy/./sympy/solvers/solvers.py", line 2979, in nsolve
  File "/Users/aaronmeurer/anaconda3/lib/python3.9/site-packages/mpmath/calculus/optimization.py", line 969, in findroot
    for x, error in iterations:
  File "/Users/aaronmeurer/anaconda3/lib/python3.9/site-packages/mpmath/calculus/optimization.py", line 93, in __iter__
    f1 = f(x1)
  File "<lambdifygenerated-55>", line 2, in _lambdifygenerated
    return (x**4 - 4*x**3 + 6*x**2 - 4*x + 1)/(4*x**3 - 12*x**2 + 12*x - 4)
  File "<string>", line 7, in __div__
  File "/Users/aaronmeurer/anaconda3/lib/python3.9/site-packages/mpmath/libmp/libmpf.py", line 956, in mpf_div
    if t == fzero: raise ZeroDivisionError
ZeroDivisionError
```

In this case, using a different solver may produce better results, like `'mnewton'` which works better for multiple roots (mnewton also handles the cos(x) = 1 case well, so I'm unsure if it's ever better to use f/f' rather than just switching to mnewton).

For removing large constants, I agree that can be useful, especially if you can't get convergence, although the recipe you give might not always be correct (e.g., it might instead require doing a change of variables).

Author: Removed this section--I think it'd get too in-the-weeds for a user-friendly guide.
